### PR TITLE
Cleanup volatile db

### PR DESCRIPTION
**Pull request by kderme**
Addresses most parts of https://github.com/input-output-hk/ouroboros-network/issues/1119

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 1]
```diff
@@ -428,25 +441,16 @@ nextFile :: forall h m blockId. IOLike m
...
-            let file = filePath $ _nextNewFileId st
-            hndl <- hOpen file (AppendMode MustBeNew)
-            return $ st {
-                  _currentWriteHandle = hndl
-                , _currentWritePath   = file
-                , _currentWriteOffset = 0
-                , _nextNewFileId      = (_nextNewFileId st) + 1
-                , _currentMap         = Map.insert file FileInfo.empty (_currentMap st)
-            }
-        (file, size) : rest -> do
```
Comment (kderme) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: This was basically dead code, so the _nextWriteFiles field is totally erased. In the past, we supported a number of policies related to files which where not full: We could for example first try to fill these files and then proceed to creating new ones. The `_nextWriteFiles` field kept track of this. However we later deleted these policies and kept the default one (append to an existing non-empty file only if it is the latest in the db).

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 2]
```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
...
+            [ "Volatile db invariant violation:"
+            , "A file was found with less than "
+            , show n
+            , " blocks, but there are no files parsed."]
+
+        truncateOnError Nothing _ _ = return ()
+        truncateOnError (Just _) file offset =
+            -- the handle of the parser is closed at this point. We need
+            -- to reOpen the file in AppendMode now (parser opens with
+            -- ReadMode).
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Capitalise the sentence and make it clear what is code by using Haddock markup (`'..'` or `@..@` for non-linked things):
```haskell
-- The handle of the parser is closed at this point. We need
-- to reopen the file in 'AppendMode' now (parser opens with
-- 'ReadMode').

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 3]
```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
...
+            --
+            withFile hasFS file (AppendMode AllowExisting) $ \hndl ->
+                hTruncate hasFS hndl (fromIntegral offset)
+
         go :: Index blockId
            -> ReverseIndex blockId
            -> SuccessorsIndex blockId
            -> Maybe (blockId, SlotNo)
-           -> [(FileId, FsPath, FileSize)] -- The relative path and size of the files with less than n blocks, if any found already.
+           -> [(FileId, FsPath, FileSize)] -- Info of files with < n blocks.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Even though it is a local function, I would make it a Haddock docstring (`-- ^ ..`).

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 4]
```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
...
+                  _currentWriteHandle = hndl
+                , _currentWritePath   = fileToWrite
+                , _currentWriteOffset = offset'
+                , _nextNewFileId      = nextNewFileId'
+                , _currentMap         = mp'
+                , _currentRevMap      = revMp
+                , _currentSuccMap     = succMp
+                , _currentMaxSlotNo   = FileInfo.maxSlotInFiles (Index.elems mp')
+            }
+            where
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: `where` indentation

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 5]
```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
...
+                        (_, Just lst) ->
+                            -- If it's not the last file, we just ignore it and
+                            -- open a new one.
+                            newFileInfo mp $ lst + 1
+
+        updateAndGo mp revMp succMp maxSlot haveLessThanN file restFiles blocks mErr = do
+            truncateOnError mErr file offset
+            newRevMp <- fromEither err $ reverseMap file revMp fileMp
+            go newMp newRevMp newSuccMp newMaxSlot newHaveLessThanN restFiles
+            where
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: `where` indentation

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 6]
```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
...
+            -- ReadMode).
+            --
+            -- Note that no file is open at this point, so we can safely
+            -- open with AppendMode any file, without the fear of opening
+            -- multiple concurrent writers, which is not allowed.
+            --
+            withFile hasFS file (AppendMode AllowExisting) $ \hndl ->
+                hTruncate hasFS hndl (fromIntegral offset)
+
         go :: Index blockId
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: This looks already better than before, but could still use some improvement. Can you start out with adding simple docstrings to `go` and `updateAndGo`?

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 7]
```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
...
+                        (Just (fd, wrfile, size), Just lst) | fd == lst->
+                            -- if the file with biggest index is non-empty,
+                            -- then this is the file we're looking for.
+                            (wrfile, lst + 1, mp, size)
+                        (_, Just lst) ->
+                            -- If it's not the last file, we just ignore it and
+                            -- open a new one.
+                            newFileInfo mp $ lst + 1
+
+        updateAndGo mp revMp succMp maxSlot haveLessThanN file restFiles blocks mErr = do
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: With so many arguments, I think a type signature would help me understand it better.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 8]
```diff
@@ -611,36 +652,51 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
...
+       => VolatileDBEnv m blockId
+       -> (forall h. InternalState blockId h -> a)
+       -> m a
+getter VolatileDBEnv{..} fromSt = do
+    mSt <- readMVar _dbInternalState
+    case mSt of
+        VolatileDbClosed  -> EH.throwError _dbErr $ UserError ClosedDBError
+        VolatileDbOpen st -> return $ fromSt st
+
+getterStm :: forall m blockId a. IOLike m
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Can you rename it to `getterSTM`? That fits better with our conventions.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 9]
```diff
@@ -611,36 +652,51 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
...
+          => VolatileDBEnv m blockId
+          -> (forall h. InternalState blockId h -> a)
+          -> STM m a
+getterStm VolatileDBEnv{..} fromSt = do
+    mSt <- readMVarSTM _dbInternalState
+    case mSt of
+        VolatileDbClosed  -> EH.throwError' _dbErrSTM $ UserError ClosedDBError
+        VolatileDbOpen st -> return $ fromSt st
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: :+1:

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 10]
```diff
@@ -611,36 +652,51 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
...
 
 reverseMap :: forall blockId
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Can you add a docstring?

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 11]
```diff
@@ -142,16 +141,23 @@ volatileDbIsOpen :: OpenOrClosed blockId h -> Bool
...
-    , _currentWriteOffset :: !Word64 -- The 'WriteHandle' for the same file.
-    , _nextWriteFiles     :: ![(FsPath, Word)] -- The path and the size of the next files to write
-    , _nextNewFileId      :: !Int -- The next file name Id.
-    , _currentMap         :: !(Index blockId) -- The content of each file.
-    , _currentRevMap      :: !(ReverseIndex blockId) -- Where to find each block from slot.
-    , _currentSuccMap     :: !(SuccessorsIndex blockId) -- successors for each block.
-    , _currentMaxSlotNo   :: !MaxSlotNo -- Highest ever stored SlotNo
+data InternalState blockId h = InternalState
+    { _currentWriteHandle :: !(Handle h)
+    -- ^ The unique open file we append blocks.
```
[the remark this answers is not in the document]
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: It highlights that there is a single block we write to, at each time.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 12]
```diff
@@ -254,68 +260,84 @@ getBlockImpl env@VolatileDBEnv{..} slot = do
...
-                if not (FileInfo.isFull _maxBlocksPerFile fileInfo')
-                then return (st', ())
-                else (\s -> (s,())) <$> nextFile hasFS _dbErr env st'
+putBlockImpl env@VolatileDBEnv{..} binfo@BlockInfo{..} builder =
+  modifyState env $ \hasFS@HasFS{..} st@InternalState{..} ->
+    if Map.member bbid _currentRevMap
+    then return (st, ()) -- putting an existing block is a no-op.
+    else do
+        bytesWritten <- hPut hasFS _currentWriteHandle builder
+        putBlockImpl' env binfo hasFS st bytesWritten
```
[the remark this answers is not in the document]
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: @mrBliss I am a bit confused about 4 vs 2 spaces. in which cases, do we have to use 4 but not 2?

[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: The guideline in this repository is: use 2 spaces for indentation, unless there's a `where`-clause for the given block, in which case you should indent the `where` keyword 2 spaces and the block above the `where` keyword and the bindings below it with 4 spaces.

So:
```haskell
foo ... =
  bla di bla
```
but
```haskell
foo ... =
    bla di bla
  where
    bla = ..
```
In nested code blocks, the same principle holds.

I intentionally deviate from this guideline in one case: I indent the body of a top-level function (or a big local function) with 4 spaces by default, even without a `where`-clause, so that in case we add one, we don't have to reindent.

**Comment by kderme** [comment]
@mrBliss Thanks for the extensive review!
> It's kind of fishy that we're using `modifyState` in a getter. I see that you just return the original state, but still.
> 
> Can you create a variant of `modifyState` that is read-only? See `withOpenState` from ImmutableDB.
> I don't think you need to do the whole `StateT` thing.

I think this fits well in my next task https://github.com/input-output-hk/ouroboros-network/issues/767.

> I would like to get rid of this one (used currently in the tests). Certainly now you have `getter` and `getterStm`.

I got some errors about escaping parameter when I tried to use the new 'getter' for the 'getInternalState'. I moved it now in testing modules. I keep in mind to rewrite it if I find a better way.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 13]
```diff
@@ -244,8 +255,8 @@ getBlockImpl :: (IOLike m, Ord blockId)
              => VolatileDBEnv m blockId
              -> blockId
              -> m (Maybe ByteString)
-getBlockImpl env@VolatileDBEnv{..} slot = do
-    modifyState env $ \hasFS@HasFS{..} st@InternalState{..} -> do
+getBlockImpl env@VolatileDBEnv{..} slot =
+    modifyState env $ \hasFS@HasFS{..} st@InternalState{..} ->
```
[the remark this answers is not in the document]
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: I think this fits well in my next task https://github.com/input-output-hk/ouroboros-network/issues/767.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: I see. I think "sole" or "only" is more accurate in that case.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 14]
```diff
@@ -228,226 +239,220 @@ isOpenDBImpl VolatileDBEnv{..} = do
...
+        | not isCurrent -> do
+            removeFile $ filePath fileId
+            return st {
+                _currentMap = Index.delete fileId _currentMap
+              , _currentRevMap = currentRevMap'
+              , _currentSuccMap = succMap'
+              }
+        | isCurrentNew  -> return st
+        | otherwise     -> do
+            -- TODO: Rething if we want to reopen after
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: `s/Rething/Rethink/`

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs [hunk 15]
```diff
@@ -74,20 +75,32 @@ wrapFsError :: Monad m
...
 
--- Throws an error if one of the given file names does not parse.
+-- | Parses the 'FsPath' of each filename and zips them together.
+-- If any 'FsPath' does not parse, we return an error.
+parseAllFds :: forall blockId.
+               [FsPath]
+            -> Either (VolatileDBError blockId) [(FileId, FsPath)]
+parseAllFds = mapM go
+  where
+    go :: FsPath -> Either (VolatileDBError blockId) (FileId, FsPath)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs]: Rename `go` to `parseFd`. Two reasons: (1) `go` is mostly used for functions that do explicit recursion, which is not the case here, the recursion is handled by `mapM`; (2) we mostly use `go` when we can't think of a good name, but in this case there is a  good candidate :slightly_smiling_face:

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs [hunk 16]
```diff
@@ -74,20 +75,32 @@ wrapFsError :: Monad m
...
+    go file = case parseFd file of
+      Nothing -> Left $ UnexpectedError $ ParserError $ InvalidFilename file
+      Just fd -> Right (fd, file)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs]: You could consider using `throwError` instead of `Left` and `return` instead of `Right`. Up to you.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs [hunk 17]
```diff
@@ -74,20 +75,32 @@ wrapFsError :: Monad m
...
-        go fd file = case parseFd file of
-            Nothing  -> Left $ UnexpectedError $ ParserError $ InvalidFilename file
-            Just fd' -> Right $ Just $ maxMaybe fd fd'
+findLastFd = foldM go Nothing
+  where
+    maxMaybe :: Ord a => Maybe a -> a -> a
+    maxMaybe ma a = case ma of
+      Nothing -> a
+      Just a' -> max a' a
+    go :: Maybe FileId -> FsPath -> Either (VolatileDBError blockId) (Maybe FileId)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs]: Hmm, this is basically the same `go` as the one in `parseAllFds`. Rename it to `parseFd`, pull out the `maxMaybe`, and promote it to the top-level so it can be shared between `parseAllFds` and `findLastFd`.

Ah, I see `parseFd` already exists. I would unify the two:
```haskell
parseFd :: FsPath -> Either (VolatileDBError blockId) FileId
parseFd file = maybe err Right $
    parseFilename <=< lastMaybe $ fsPathToList file
  where
    err = Left $ UnexpectedError $ ParserError $ InvalidFilename file

    parseFilename :: Text -> Maybe FileId
    parseFilename = readMaybe
                  . T.unpack
                  . snd
                  . T.breakOnEnd "-"
                  . fst
                  . T.breakOn "."
```

Update `unsafeParseFd` by using `either .. id ..`, 
`parseAllFds = mapM $ \f -> (,f) <$> parseFd f`, and
`findLastFd = fmap safeMaximum . mapM parseFd . Set.toList`.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/TestBlock.hs [hunk 18]
```diff
@@ -154,9 +155,18 @@ createFileImpl :: IOLike m
...
 createFileImpl hasFS env = do
-    SomePair _stHasFS st <- Internal.getInternalState env
+    SomePair _stHasFS st <- getInternalState env
     let nextFd = Internal._nextNewFileId st
     let path = Internal.filePath nextFd
     withFile hasFS path (AppendMode MustBeNew) $ \_hndl -> do
         return ()
     return ()
+
+getInternalState :: forall m blockId. IOLike m
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/TestBlock.hs]: I was more hoping that we wouldn't need access to the internal state of the VolatileDB in the tests anymore.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 19]
```diff
@@ -458,120 +463,157 @@ reOpenFile :: forall m h blockId
...
+            newRevMp <- fromEither err $ reverseMap file currentRevMap fileMp
+            go newMp newRevMp newSuccMp newMaxSlot newHaveLessThanN rest
+          where
+            offset = case reverse blocks of
+              [] -> 0
+              (slotOffset, (blockSize,_)) : _ ->
+                -- The file offset is given by the offset of the last
+                -- block plus its size.
+                slotOffset + blockSize
+            fileMp = Map.fromList blocks
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Use `Map` (like the suggested in my review) instead of `Mp` throughout this function.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 20]
```diff
@@ -458,120 +463,157 @@ reOpenFile :: forall m h blockId
...
+      , newIndex + 1
+      , Index.insert newIndex FileInfo.empty mp
+      , 0 )
+
+    errorMsg = unwords
+        [ "VolatileDB invariant violation:"
+        , "The impossible happened!"
+        , "A file was found with less than "
+        , show n
+        , " blocks, but there are no files in db!"]
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: You don't need the space after "than " and before " blocks", `unwords` is taking care of them already.

"the" missing before "db!".

I still don't really understand the message: does it mean that a file was found (with less than x blocks?) while we didn't expect to find one?

Also, can you put it closer to where it is used?

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 21]
```diff
@@ -228,226 +239,220 @@ isOpenDBImpl VolatileDBEnv{..} = do
     mSt <- readMVar _dbInternalState
     return $ volatileDbIsOpen mSt
 
--- closeDB . reOpenDB is a no-op. This is achieved because when we reOpen
--- we try to append on the latest created file.
+-- | Property: @'closeDB' >> 'reOpenDB'@ is a no-op. This is true because
+-- 'reOpenDB` will always append to the last created file.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: ` should be '

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 22]
```diff
@@ -458,120 +463,157 @@ reOpenFile :: forall m h blockId
...
+        [ "VolatileDB invariant violation:"
+        , "The impossible happened!"
+        , "A file was found with less than "
+        , show n
+        , " blocks, but there are no files in db!"]
+
+    truncateOnError Nothing _ _ = return ()
+    truncateOnError (Just _) file offset =
+        -- The handle of the parser is closed at this point. We need
+        -- to reOpen the file in 'AppendMode' now (parser opens with
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: `s/reOpen/reopen/`, because `reOpen` makes me think it is referring to a function, but it is not.

**Comment by mrBliss** [comment]
Also, can you fix the warnings produced by `-Widentities`? Only those in the VolatileDB.

**Comment by kderme** [comment]
`withState` and cleaning tests (which includes the `getInternalState`) come in a next pr.

I think I will squash all commits of this pr to a single "VolatileDB cleanup". Not sure if I can do something better at this point.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs (hunk 20 above)
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Eventually, I changed a bit the implementation and it is not needed any more.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 23]
```diff
@@ -146,9 +146,9 @@ volatileDbIsOpen :: OpenOrClosed blockId h -> Bool
 volatileDbIsOpen (VolatileDbOpen _) = True
 volatileDbIsOpen VolatileDbClosed   = False
 
-data InternalState blockId h = InternalState
-    { _currentWriteHandle :: !(Handle h)
-      -- ^ The unique open file we append blocks.
+data InternalState blockId h = InternalState {
+      _currentWriteHandle :: !(Handle h)
+      -- ^ The only open file we append blocks.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: "we append blocks *to*" or "*to which* we append blocks".

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 24]
```diff
@@ -374,7 +374,12 @@ tryCollectFile hasFS@HasFS{..} env slot st@InternalState{..} (fileId, fileInfo)
               }
         | isCurrentNew  -> return st
         | otherwise     -> do
-            -- TODO: Rething if we want to reopen after
+            -- We reach this case if we have to garbage collect the current file
+            -- we are appending blocks to. I'm not sure how realistic this case
+            -- is and maybe we want to skip this garbage collection.
+            --
+            -- 'ReOpenFile' techinally truncates the file to 0 offset, so any
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: `s/ReOpenFile/reOpenFile/` `s/techinally/technically/`

Ah, I now understand this case.

> I'm not sure how realistic this case is and maybe we want to skip this garbage collection.

Can you verify that this case is actually triggered in the tests? You can simply let it call `error` and see whether it is thrown. Maybe you'll need to play with the test parameters.

> I'm not sure how realistic this case is and maybe we want to skip this garbage collection.

I would replace this with an explanation:

> For this to happen, a garbage collection would have to be triggered for a slot ...

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 25]
```diff
@@ -479,21 +478,20 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
...
+    dbDir = mkFsPath []
 
     toFsPath :: String -> FsPath
     toFsPath file = mkFsPath [file]
 
 -- | Makes the 'InternalState' by parsing all files.
 --
--- This function may create a new file to append new blocks or use an existing
--- one.
+-- It may create a new file to append new blocks or use an existing one.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: new blocks *to*

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 26]
```diff
@@ -506,33 +504,26 @@ mkInternalState hasFS err parser n files =
...
-      , newIndex + 1
       , Index.insert newIndex FileInfo.empty mp
       , 0 )
 
-    errorMsg = unwords
-        [ "VolatileDB invariant violation:"
-        , "The impossible happened!"
-        , "A file was found with less than "
-        , show n
-        , " blocks, but there are no files in db!"]
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Ah, nice that this is no longer necessary.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 27]
```diff
@@ -506,33 +504,26 @@ mkInternalState hasFS err parser n files =
...
-        -- 'ReadMode').
-        --
-        -- Note that no file is open at this point, so we can safely
-        -- open with 'AppendMode' any file, without the fear of opening
-        -- multiple concurrent writers, which is not allowed.
-        --
-        withFile hasFS file (AppendMode AllowExisting) $ \hndl ->
-            hTruncate hasFS hndl (fromIntegral offset)
+      -- The handle of the parser is closed at this point. We need
+      -- to reοpen the file in 'AppendMode' now (parser opens with
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: The `ο` in reοpen is a o micron, replace it with an ASCII o :slightly_smiling_face:

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs [hunk 28]
```diff
@@ -75,32 +78,23 @@ wrapFsError :: Monad m
 wrapFsError fsErr volDBErr action =
     tryVolDB fsErr volDBErr action >>= either (throwError volDBErr) return
 
--- | Parses the 'FsPath' of each filename and zips them together.
--- If any 'FsPath' does not parse, we return an error.
+-- | Parses the 'FileId' of each 'FsPath' and zips them together.
+-- If any parsing does not parse, we abort with an error.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs]: `If any parsing does not parse, we abort with an error.` -> `When parsing fails, we abort with the corresponding parse error.`

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs [hunk 29]
```diff
@@ -75,32 +78,23 @@ wrapFsError :: Monad m
...
-    maxMaybe ma a = case ma of
-      Nothing -> a
-      Just a' -> max a' a
-    go :: Maybe FileId -> FsPath -> Either (VolatileDBError blockId) (Maybe FileId)
-    go fd file = case parseFd file of
-      Nothing  -> Left $ UnexpectedError $ ParserError $ InvalidFilename file
-      Just fd' -> Right $ Just $ maxMaybe fd fd'
+findLastFd = fmap safeMaximum . mapM parseFd . Set.toList
+
+maxMaybe :: Ord a => Maybe a -> a -> a
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs]: Is this function still needed? I'd like to get rid of it. Use `safeMaximum` instead.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs [hunk 30]
```diff
@@ -117,12 +111,12 @@ tryVolDB :: forall m blockId a. Monad m
...
-    where
-      fromFS = UnexpectedError . FileSystemError
+  where
+    fromFS = UnexpectedError . FileSystemError
 
-      squash :: Either FsError (Either (VolatileDBError blockId) a)
-               -> Either (VolatileDBError blockId) a
-      squash = either (Left . fromFS) id
+    squash :: Either FsError (Either (VolatileDBError blockId) a)
+              -> Either (VolatileDBError blockId) a
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs]: Please align the `->` with the `::`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/Model.hs [hunk 31]
```diff
@@ -49,7 +49,7 @@ import           Ouroboros.Storage.Util.ErrorHandling (ThrowCantCatch)
 import qualified Ouroboros.Storage.Util.ErrorHandling as EH
 import           Ouroboros.Storage.VolatileDB.API
 import qualified Ouroboros.Storage.VolatileDB.Impl as Internal
-import           Ouroboros.Storage.VolatileDB.Util
+import           Ouroboros.Storage.VolatileDB.Util hiding (maxMaybe)
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/Model.hs]: So `maxMaybe` is defined here too? Replace it with `safeMaximum` too. If necessary use `max`, which works with two `Maybe a`s.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs [hunk 32]
```diff
@@ -2,12 +2,12 @@
 {-# LANGUAGE LambdaCase          #-}
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs]: Please add an export list to this module.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs (hunk 27 above)
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: :man_facepalming: :man_facepalming:

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 33]
```diff
@@ -611,36 +649,46 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
...
+getterSTM fromSt VolatileDBEnv{..} = do
+    mSt <- readMVarSTM _dbInternalState
+    case mSt of
+      VolatileDbClosed  -> EH.throwError' _dbErrSTM $ UserError ClosedDBError
+      VolatileDbOpen st -> return $ fromSt st
 
-reverseMap :: forall blockId
-           .  Ord blockId
+-- | For each block found in a parsed file, we insert its 'InternalBlockInfo'.
+-- If the block is already found in the 'ReverseIndex' or is dublicated in the
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: `s/dublicated/duplicated/`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/TestBlock.hs [hunk 34]
```diff
@@ -75,17 +76,17 @@ myParser hasFs = Parser {
 parseImpl :: forall m h. (MonadThrow m)
           => HasFS m h
           -> FsPath
-          -> m ([(SlotOffset, (BlockSize, BlockInfo BlockId))], Maybe String)
+          -> m ([(SlotOffset, (Word64, BlockInfo BlockId))], Maybe String)
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/TestBlock.hs]: Why this change? `BlockSize` is clearer than `Word64` to me.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 35]
```diff
@@ -228,226 +240,221 @@ isOpenDBImpl VolatileDBEnv{..} = do
     mSt <- readMVar _dbInternalState
     return $ volatileDbIsOpen mSt
 
--- closeDB . reOpenDB is a no-op. This is achieved because when we reOpen
--- we try to append on the latest created file.
+-- | Property: @'closeDB' >> 'reOpenDB'@  should be a no-op. This is true
+-- because 'reOpenDB` will always append to the last created file.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]:
```suggestion
-- because 'reOpenDB' will always append to the last created file.
```

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs (hunk 24 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: > Can you verify that this case is actually triggered in the tests? You can simply let it call error and see whether it is thrown. Maybe you'll need to play with the test parameters.

@kderme Have you seen this happen?

Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Yes, q-s-m triggers this case. I left the comment:
```
-- We reach this case if we have to garbage collect the current file 
-- we are appending blocks to. For this to happen, a garbage
-- collection would have to be triggered for a slot which is bigger
-- than any recently inserted blocks.
```

**Comment by kderme** [comment]
@mrBliss Yes q-s-m triggers this case frequently.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/TestBlock.hs (hunk 34 above)
Comment (mrBliss) [reply ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/TestBlock.hs]: I see, ok.

**Comment by mrBliss** [comment]
If you squash your commits titled "Address review" (into the first one, I suppose), we can merge this!

**Comment by mrBliss** [comment]
> If you squash your commits titled "Address review" (into the first one, I suppose), we can merge this!

I meant: squash all *your* commits titled "Address review" into the very first commit ("cleanup Volatile db"). Or, if some of the changes in the "Address review" commits belong to other commits in those.

I don't think that "Address review" commits in general belong in master, we don't need to see the whole history of each PR.

**Comment by kderme** [comment]
@mrBliss I see, I squashed now my commits properly. 

I also checked the diff of the branches before and after the rebase and they are almost identical except for some small improvements.

**Comment by mrBliss** [comment]
Thanks @kderme!
